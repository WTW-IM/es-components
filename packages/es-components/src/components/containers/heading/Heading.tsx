import React from 'react';
import PropTypes from 'prop-types';
import styled, {
  DefaultTheme,
  ThemedStyledProps,
  css
} from 'styled-components';
import { HeadingLevel } from 'es-components-shared-types';

const knockoutStyles = css`
  background-color: ${({ theme }) => theme.colors.primary};
  color: white;
  padding: 20px 15px;
`;

export type HeadingProps = JSXElementProps<'h1'> & {
  children?: React.ReactNode;
  level?: HeadingLevel;
  size?: HeadingLevel;
  isKnockoutStyle?: boolean;
  underlineColor?: string | null;
};

function getAdjustedProps(
  func: (
    props: ThemedStyledProps<HeadingProps, DefaultTheme> & {
      adjustedSize: HeadingLevel;
    }
  ) => string
) {
  return ({
    size,
    level = 1,
    ...props
  }: ThemedStyledProps<HeadingProps, DefaultTheme>) =>
    func({
      ...props,
      size,
      level,
      adjustedSize: size || level
    });
}

const UnstyledHeading = styled.h1``;
const Heading = styled(({ level = 1, ...props }: HeadingProps) => (
  <UnstyledHeading as={`h${level}`} {...props} />
))`
  border-bottom: ${props =>
    props.underlineColor && `2px solid ${props.underlineColor};`};
  color: inherit;
  font-family: 'Source Sans Pro', 'Segoe UI', Segoe, Calibri, Tahoma, sans-serif;
  font-size: ${getAdjustedProps(
    ({ theme, adjustedSize }) => theme.font.headingMobile[adjustedSize]
  )};
  font-weight: 300;
  line-height: 1.1;
  margin-bottom: 0.45em;
  margin-top: 0;
  padding-bottom: ${props => props.underlineColor && '0.22em'};

  && {
    ${({ isKnockoutStyle }) => isKnockoutStyle && knockoutStyles}
  }

  small {
    font-size: ${getAdjustedProps(({ adjustedSize }) =>
      adjustedSize > 3 ? '75%' : '65%'
    )};
    line-height: 1;
  }

  @media (min-width: ${props => props.theme.screenSize.tablet}) {
    font-size: ${getAdjustedProps(
      ({ adjustedSize, theme }) => theme.font.headingDesktop[adjustedSize]
    )};
  }
`;

Heading.propTypes = {
  children: PropTypes.node,
  /** Heading level element */
  level: PropTypes.oneOf([1, 2, 3, 4, 5, 6]),
  /** Override the default font size with another level */
  size: PropTypes.oneOf([1, 2, 3, 4, 5, 6]),
  /** Alternate knockout-style header with a solid background */
  isKnockoutStyle: PropTypes.bool,
  /** Include an underline with the provided color */
  underlineColor: PropTypes.string
};

Heading.defaultProps = {
  children: undefined,
  level: 1,
  size: undefined,
  isKnockoutStyle: false,
  underlineColor: null
};

/** @component */
export default Heading;

export const PageHeading = styled(Heading).attrs({
  isKnockoutStyle: true,
  level: 1
})``;
