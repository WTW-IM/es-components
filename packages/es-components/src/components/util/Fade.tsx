import React from 'react';
import PropTypes from 'prop-types';
import Transition, {
  TransitionStatus,
  TimeoutProps
} from 'react-transition-group/Transition';

type Opacity = React.CSSProperties['opacity'];

interface FadeProps extends Partial<TimeoutProps<HTMLElement>> {
  duration?: number;
  opacity?: Opacity;
  children: NonNullable<React.ReactNode>;
}

const getFadeStyle = (status: TransitionStatus, finalOpacity: Opacity) => {
  switch (status) {
    case 'entering':
    case 'entered':
      return { finalOpacity };
    default:
      return { opacity: 0 };
  }
};

type AnyReactHTMLElement = React.ReactHTMLElement<HTMLElement>;

const TransitionChild: React.FC<{
  children: AnyReactHTMLElement;
  style: React.CSSProperties;
  transitionStatus: TransitionStatus;
}> = ({ children, style, transitionStatus }) => {
  return React.cloneElement(children, {
    style: {
      ...((children?.props?.style as Maybe<React.CSSProperties>) || {}),
      ...style,
      ...getFadeStyle(
        transitionStatus,
        typeof style.opacity == 'number' || Boolean(style.opacity)
          ? style.opacity
          : 1
      )
    }
  });
};

const Fade: React.FC<FadeProps> = ({
  children,
  duration,
  opacity,
  ...otherProps
}) => {
  const transitionStyles = {
    transition: `opacity ${duration || 150}ms linear`,
    opacity
  };

  const transitionProps = {
    ...otherProps,
    timeout: duration || otherProps.timeout || 150
  };

  const hasValidChildren =
    children &&
    (typeof children === 'object' || typeof children === 'function');

  return !hasValidChildren ? (
    <>{children}</>
  ) : (
    <Transition {...transitionProps}>
      {transitionStatus => {
        try {
          return (
            React.Children.only(children) && (
              <TransitionChild
                transitionStatus={transitionStatus}
                style={transitionStyles}
              >
                {children as AnyReactHTMLElement}
              </TransitionChild>
            )
          );
        } catch (e) {
          return React.Children.map(children, child => {
            return !child ? (
              child
            ) : (
              <TransitionChild
                transitionStatus={transitionStatus}
                style={transitionStyles}
              >
                {child as AnyReactHTMLElement}
              </TransitionChild>
            );
          });
        }
      }}
    </Transition>
  );
};

Fade.propTypes = {
  children: PropTypes.any.isRequired,
  duration: PropTypes.number,
  opacity: PropTypes.number
};

Fade.defaultProps = {
  duration: 150,
  opacity: 1
};

export default Fade;
