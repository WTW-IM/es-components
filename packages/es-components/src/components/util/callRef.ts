import React from 'react';

const callRef = <T, R extends React.ForwardedRef<T>>(ref: Maybe<R>, value: T) =>
  typeof ref === 'function'
    ? ref(value)
    : ref && Object.hasOwn(ref, 'current') && (ref.current = value);

export default callRef;

export function callRefs<T, R extends React.ForwardedRef<T>>(
  value: T,
  ...refs: R[]
) {
  refs.forEach(r => callRef(r, value));
}
