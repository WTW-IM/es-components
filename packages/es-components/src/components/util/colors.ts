import tinycolor, { ColorInput } from 'tinycolor2';

export function darken(color: ColorInput, percent: number) {
  return tinycolor(color).darken(percent).toHexString();
}

export function lighten(color: ColorInput, percent: number) {
  return tinycolor(color).lighten(percent).toHexString();
}

export function getTextColor(
  bgColor: ColorInput,
  darkestColor: ColorInput = 'black',
  lightestColor: ColorInput = 'white'
) {
  return tinycolor.readability(darkestColor, bgColor) > 10
    ? darkestColor
    : lightestColor;
}
