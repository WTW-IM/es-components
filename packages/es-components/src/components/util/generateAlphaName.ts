export default function generateAlphaName(): string {
  const possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz';
  let text = '';
  for (let i = 0; i < 6; i++) {
    text += possible.charAt(Math.floor(Math.random() * possible.length));
  }
  return text;
}
