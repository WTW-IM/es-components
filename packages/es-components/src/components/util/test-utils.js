// eslint-env jest
import React from 'react';
import PropTypes from 'prop-types';
import { ThemeProvider } from 'styled-components';
import viaTheme from 'es-components-via-theme';
import { render } from '@testing-library/react';

export function ThemeComponent(props) {
  return <ThemeProvider theme={viaTheme} {...props} />;
}

ThemeComponent.propTypes = {
  children: PropTypes.node
};

export function renderWithTheme(component) {
  return render(<ThemeComponent>{component}</ThemeComponent>);
}
