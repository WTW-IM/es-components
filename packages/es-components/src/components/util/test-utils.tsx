import React, { Profiler } from 'react';
import PropTypes from 'prop-types';
import { ThemeProvider } from 'styled-components';
import viaTheme from 'es-components-via-theme';
import { render, RenderResult } from '@testing-library/react';

export function ThemeComponent(
  props: Omit<React.ComponentPropsWithoutRef<typeof ThemeProvider>, 'theme'>
) {
  return <ThemeProvider {...props} theme={viaTheme} />;
}

ThemeComponent.propTypes = {
  children: PropTypes.node
};

const profilerRender = (id: string, phase: string, actualDuration: number) => {
  if (actualDuration < 100) return;
  console.warn('Render took too long!', { id, phase, actualDuration });
};

export function renderWithTheme(component: React.ReactElement): RenderResult {
  const { ...viewUtils } = render(<ThemeComponent>{component}</ThemeComponent>);
  return {
    ...viewUtils,
    rerender: (rerenderComponent: React.ReactNode) =>
      viewUtils.rerender(
        <Profiler id="ThemeRender" onRender={profilerRender}>
          <ThemeComponent>{rerenderComponent}</ThemeComponent>
        </Profiler>
      )
  };
}

export const setClientWidth = (width: number) => {
  Object.defineProperty(document.documentElement, 'clientWidth', {
    writable: true,
    configurable: true,
    value: width
  });
};

export const setClientHeight = (height: number) => {
  Object.defineProperty(document.documentElement, 'clientHeight', {
    writable: true,
    configurable: true,
    value: height
  });
};
