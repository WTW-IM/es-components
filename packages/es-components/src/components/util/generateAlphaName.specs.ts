import generateAlphaName from './generateAlphaName';

describe('generateAlphaName', () => {
  it('generates a 6 character string', () => {
    const result = generateAlphaName();

    expect(result.length).toBe(6);
  });

  it('generates only alpha characters', () => {
    const result = generateAlphaName();

    expect(/^[a-zA-Z]+$/.test(result)).toBe(true);
  });

  it('generates a unique string each subsequent execution', () => {
    const result1 = generateAlphaName();
    const result2 = generateAlphaName();

    expect(result1).not.toEqual(result2);
  });
});
