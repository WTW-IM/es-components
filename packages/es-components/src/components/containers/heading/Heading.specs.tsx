/* eslint-env jest */
import React from 'react';
import styled from 'styled-components';
import viaTheme from 'es-components-via-theme';
import { renderWithTheme } from '../../util/test-utils';
import Heading from './Heading';
import { screen } from '@testing-library/react';

it('renders proper heading level', () => {
  renderWithTheme(<Heading level={1}>Heading level 1</Heading>);

  expect(screen.getByText('Heading level 1')).not.toBeNull();
});

it('renders heading level with another size', () => {
  const { container } = renderWithTheme(
    <div>
      <Heading level={1}>Heading level 1</Heading>
      <Heading level={3}>Heading level 3</Heading>
      <Heading level={3} size={1}>
        Heading level 3 size 1
      </Heading>
    </div>
  );

  expect(container).toMatchSnapshot();
});

it('renders knockout heading with different class', () => {
  const { container } = renderWithTheme(
    <div>
      <Heading level={1}>Heading level 1</Heading>
      <Heading level={1} isKnockoutStyle>
        Knockout Heading level 1
      </Heading>
    </div>
  );

  expect(container).toMatchSnapshot();
});

it('works properly with downstream attrs', async () => {
  const TestHeading = styled(Heading).attrs(() => ({ level: 2, size: 4 }))``;
  renderWithTheme(<TestHeading>This is a test.</TestHeading>);

  const foundTitle = await screen.findByRole('heading');
  expect(foundTitle).toBeInTheDocument();
  expect(foundTitle.tagName.toLowerCase()).toBe('h2');
  expect(foundTitle).toHaveStyleRule(
    'font-size',
    viaTheme.font.headingMobile[4]
  );
});
