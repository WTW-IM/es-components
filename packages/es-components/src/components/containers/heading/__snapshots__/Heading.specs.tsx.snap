// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders heading level with another size 1`] = `
.c0 {
  color: inherit;
  font-family: 'Source Sans Pro','Segoe UI',Segoe,Calibri,Tahoma,sans-serif;
  font-size: 38.7898px;
  font-weight: 300;
  line-height: 1.1;
  margin-bottom: 0.45em;
  margin-top: 0;
}

.c0 small {
  font-size: 65%;
  line-height: 1;
}

.c1 {
  color: inherit;
  font-family: 'Source Sans Pro','Segoe UI',Segoe,Calibri,Tahoma,sans-serif;
  font-size: 31.104px;
  font-weight: 300;
  line-height: 1.1;
  margin-bottom: 0.45em;
  margin-top: 0;
}

.c1 small {
  font-size: 65%;
  line-height: 1;
}

@media (min-width:768px) {
  .c0 {
    font-size: 44.78976px;
  }
}

@media (min-width:768px) {
  .c1 {
    font-size: 31.104px;
  }
}

<div>
  <div>
    <h1
      class="c0"
    >
      Heading level 1
    </h1>
    <h3
      class="c1"
    >
      Heading level 3
    </h3>
    <h3
      class="c0"
      size="1"
    >
      Heading level 3 size 1
    </h3>
  </div>
</div>
`;

exports[`renders knockout heading with different class 1`] = `
.c0 {
  color: inherit;
  font-family: 'Source Sans Pro','Segoe UI',Segoe,Calibri,Tahoma,sans-serif;
  font-size: 38.7898px;
  font-weight: 300;
  line-height: 1.1;
  margin-bottom: 0.45em;
  margin-top: 0;
}

.c0 small {
  font-size: 65%;
  line-height: 1;
}

.c1 {
  color: inherit;
  font-family: 'Source Sans Pro','Segoe UI',Segoe,Calibri,Tahoma,sans-serif;
  font-size: 38.7898px;
  font-weight: 300;
  line-height: 1.1;
  margin-bottom: 0.45em;
  margin-top: 0;
}

.c1.c1 {
  background-color: #0073b6;
  color: white;
  padding: 20px 15px;
}

.c1 small {
  font-size: 65%;
  line-height: 1;
}

@media (min-width:768px) {
  .c0 {
    font-size: 44.78976px;
  }
}

@media (min-width:768px) {
  .c1 {
    font-size: 44.78976px;
  }
}

<div>
  <div>
    <h1
      class="c0"
    >
      Heading level 1
    </h1>
    <h1
      class="c1"
    >
      Knockout Heading level 1
    </h1>
  </div>
</div>
`;
